File: worker.go
package main

import (
	"bytes"
	"database/sql"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/lib/pq"
)

type Entry struct {
	Id               int
	Data             sql.NullString
	Result           sql.NullString
	Status           sql.NullString
	WorkerName       sql.NullString
	StatusChangeTime sql.NullString
}

/*
	1. Listen for channel events
	2. Use the id from channel and select the table to get new job
	3. do the work and perform an update
*/
func main() {
	logger := log.New(os.Stdout, "", log.Ldate|log.Ltime)

	//db, err := sql.Open("postgres", "user=drio password= dbname=pub_sub_test sslmode=disable")
	dsn := "postgresql://localhost:5432/pub_sub_test?sslmode=disable"
	db, err := sql.Open("postgres", dsn)
	if err != nil {
		logger.Fatal(err)
	}
	defer db.Close()
	logger.Printf("database connection pool established")

	entries := make([]*Entry, 0)
	rows, err := db.Query("SELECT * from ps_jobs")
	if err != nil {
		logger.Fatal(err)
	}
	defer rows.Close()

	for rows.Next() {
		e := new(Entry)
		if err := rows.Scan(&e.Id, &e.Data, &e.Result, &e.Status, &e.WorkerName, &e.StatusChangeTime); err != nil {
			panic(err)
		}
		entries = append(entries, e)
		fmt.Println(e.Id, e.Data.String)
	}
	if err := rows.Err(); err != nil {
		panic(err)
	}

	/* Test LISTEN */
	reportProblem := func(ev pq.ListenerEventType, err error) {
		if err != nil {
			fmt.Println(err.Error())
		}
	}

	listener := pq.NewListener(dsn, 10*time.Second, time.Minute, reportProblem)
	chName := "jobs_status_channel"
	err = listener.Listen(chName)
	if err != nil {
		panic(err)
	}

	fmt.Println("Start monitoring PostgreSQL...")
	waitForNotification(listener)
	fmt.Println("Done")
}

func waitForNotification(l *pq.Listener) {
	for {
		select {
		case n := <-l.Notify:
			fmt.Println("Received data from channel [", n.Channel, "] :")
			// Prepare notification payload for pretty print
			var prettyJSON bytes.Buffer
			err := json.Indent(&prettyJSON, []byte(n.Extra), "", "\t")
			if err != nil {
				fmt.Println("Error processing JSON: ", err)
				return
			}
			id := string(prettyJSON.Bytes())
			fmt.Println("Id is : ", id)
			return
		case <-time.After(90 * time.Second):
			fmt.Println("Received no events for 90 seconds, checking connection")
			go func() {
				l.Ping()
			}()
			return
		}
	}
}

File: worker.py
import select
import sys
import psycopg2
import psycopg2.extensions

DSN="dbname='pub_sub_test' user='drio' host='localhost' password='dbpass'"
channel = "jobs_status_channel"
table = "ps_jobs"
worker_name = None

if len(sys.argv) != 2:
    print("worker name provided")
    sys.exit(1)
worker_name = sys.argv[1]

def log(msg):
    print(f'[{worker_name}] >> {msg}')

'''
FIXME: In a real world we will want to have an intermediate step
where we set the job to 'processing'
new -> processing -> success
'''
def doWork(conn, id):
    curs = conn.cursor()
    curs.execute(f'select * from {table} where id = %s;', (id,))
    rows = curs.fetchall()
    if len(rows) == 1:
        data = rows[0][1]
        status = rows[0][3]
        if status == 'new':
            addedNumbers = sum([int(c) for c in list(data) if c.isdigit()])
            log(f"Updating id={id} with result={addedNumbers}")
            curs.execute(f"""
                UPDATE
                    {table}
                SET
                    result = %s,
                    status = %s,
                    worker_name = %s,
                    status_change_time = current_timestamp
                WHERE
                    id = %s
            """, (addedNumbers, 'success', worker_name, id))
        else:
            log(f'No new job. Nothing to do.')
    curs.close()


def listen(conn):
    curs = conn.cursor()
    curs.execute(f'LISTEN {channel};')
    log(f"Waiting for notifications on channel '{channel}'")
    while True:
        if select.select([conn],[],[],5) == ([],[],[]):
            log("Timeout")
        else:
            conn.poll()
            while conn.notifies:
                notify = conn.notifies.pop(0)
                log(f"Got NOTIFY: {notify.pid}, {notify.channel}, {notify.payload}")
                doWork(conn, notify.payload)


if __name__ == '__main__':
    conn = psycopg2.connect(DSN)
    conn.set_isolation_level(psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT)
    listen(conn)
